lua/feline/statusline_ffi.lua

```diff
@@ -0,0 +1,63 @@
+-- This module provides an interface to Neovim's statusline generator using Lua FFI
+local M = {}
+
+local api = vim.api
+local ffi = require('ffi')
+
+-- Definitions required to use Neovim's build_stl_str_hl function to expand statusline expressions
+-- And also other definitions used by this module
+ffi.cdef [[
+typedef unsigned char char8_t;
+typedef struct window_S win_T;
+extern win_T *curwin;
+typedef struct {} stl_hlrec_t;
+typedef struct {} StlClickDefinition;
+typedef struct {} StlClickRecord;
+int build_stl_str_hl(
+    win_T *wp,
+    char8_t *out,
+    size_t outlen,
+    char8_t *fmt,
+    int use_sandbox,
+    char8_t fillchar,
+    int maxwidth,
+    stl_hlrec_t **hltab,
+    StlClickRecord **tabtab
+);
+
+size_t strlen(const char* str);
+]]
+
+-- Used CType values stored in a local variable to avoid redefining them and improve performance
+local char8_buf_t = ffi.typeof('char8_t[?]')
+local char8_str_t = ffi.typeof('char8_t*')
+
+-- Expand statusline expression, returns a Lua string containing plaintext with only the characters
+-- that'll be displayed in the statusline
+function M.expand_statusline_expr(expr, winid)
+    local stlbuf = char8_buf_t(256)
+    expr = ffi.cast(char8_str_t, expr)
+
+    ffi.C.build_stl_str_hl(
+        ffi.C.curwin,
```

Comment:
Since you're expanding for winid you need to pass win_T pointer for window of that winid else you'll get wrong results.

Reply:
> Since you're expanding for winid you need to pass win_T pointer for window of that winid else you'll get wrong results.

How would I go about doing that? Sorry, I'm really not familiar with Neovim's codebase

Reply:
I figured it out. Though it'd be nice if I could reuse the window value to access its `w_width` member to avoid calling `api.nvim_win_get_width()`, but trying to do that makes Lua complain about the unknown size of the win_T struct.

Reply:
You need to put `struct window_S` definition in `cdef` if you want to access `w_width` .

Reply:
> You need to put `struct window_S` definition in `cdef` if you want to access `w_width` .

Yeah, I'm aware of that. But honestly that seems like more trouble than it's worth